### ISSUE TITLE
Error occured while extracting the APK.

### ISSUE THREAD

**Issue by aliyuchang33** [sensepost/objection#56]
```
root@xxx:~# objection patchapk -s com.mimikko.mimikkoui_1.8.6_29.apk --architecture arm64-v8a
Using latest Github gadget version: 10.6.54
Patcher will be using Gadget version: 10.6.54
Unpacking com.mimikko.mimikkoui_1.8.6_29.apk
An error may have occured while extracting the APK.
Exception in thread "main" java.lang.NullPointerException
	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1792)
	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1769)
	at org.apache.commons.io.IOUtils.copy(IOUtils.java:1744)
	at brut.androlib.res.AndrolibResources.getFrameworkApk(AndrolibResources.java:570)
	at brut.androlib.res.AndrolibResources.loadFrameworkPkg(AndrolibResources.java:112)
	at brut.androlib.res.data.ResTable.getPackage(ResTable.java:82)
	at brut.androlib.res.data.ResTable.getResSpec(ResTable.java:65)
	at brut.androlib.res.data.ResTable.getResSpec(ResTable.java:61)
	at brut.androlib.res.decoder.ResAttrDecoder.decode(ResAttrDecoder.java:39)
	at brut.androlib.res.decoder.AXmlResourceParser.getAttributeValue(AXmlResourceParser.java:369)
	at org.xmlpull.v1.wrapper.classic.XmlPullParserDelegate.getAttributeValue(XmlPullParserDelegate.java:69)
	at brut.androlib.res.decoder.XmlPullStreamDecoder$1.parseManifest(XmlPullStreamDecoder.java:97)
	at brut.androlib.res.decoder.XmlPullStreamDecoder$1.event(XmlPullStreamDecoder.java:65)
	at brut.androlib.res.decoder.XmlPullStreamDecoder.decode(XmlPullStreamDecoder.java:141)
	at brut.androlib.res.decoder.XmlPullStreamDecoder.decodeManifest(XmlPullStreamDecoder.java:153)
	at brut.androlib.res.decoder.ResFileDecoder.decodeManifest(ResFileDecoder.java:140)
	at brut.androlib.res.AndrolibResources.decodeManifestWithResources(AndrolibResources.java:199)
	at brut.androlib.Androlib.decodeManifestWithResources(Androlib.java:140)
	at brut.androlib.ApkDecoder.decode(ApkDecoder.java:100)
	at brut.apktool.Main.cmdDecode(Main.java:165)
	at brut.apktool.Main.main(Main.java:81)

App already has android.permission.INTERNET
Smali not found in smali directory. This might be a multidex APK. Searching...
Traceback (most recent call last):
  File "/usr/local/bin/objection", line 11, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/objection/console/cli.py", line 219, in patchapk
    patch_android_apk(**locals())
  File "/usr/local/lib/python3.5/dist-packages/objection/commands/mobile_packages.py", line 155, in patch_android_apk
    patcher.inject_load_library()
  File "/usr/local/lib/python3.5/dist-packages/objection/utils/patchers/android.py", line 496, in inject_load_library
    raise Exception('Unable to find smali to patch!')
Exception: Unable to find smali to patch!
Cleaning up temp files...
Failed to cleanup with error: [Errno 2] No such file or directory: '/tmp/tmpqavggvwl.apktemp.objection.apk
```

I am not sure what happened to this app. Do you need this apk?

**Comment by leonjza** [comment]
@aliyuchang33 did you come right?

**Comment by aliyuchang33** [comment]
Still the same problem. However I don't have the source of this apk so I give up.

**Comment by leonjza** [comment]
Can you share the APK with me?

**Comment by aliyuchang33** [comment]
Can you give me your email address?

**Comment by leonjza** [comment]
Sure. You can reach me at leon @ sensepost . com

**Comment by leonjza** [comment]
Got it thanks. I am also failing to patch it with the first try, but for a different reason. Will test it and let you know if I can come up with something that works. It looks like this `AndroidManifest.xml` might need a manual edit before repackaging too.

**Comment by aliyuchang33** [comment]
Thank you so much. I am learning Decompiling. Hope you can achieve that.

发件人: Leon Jacobs
发送时间: 2018年2月21日 19:44
收件人: sensepost/objection
抄送: aliyuchang33; Mention
主题: Re: [sensepost/objection] Error occured while extracting the APK.(#56)

Got it thanks. I am also failing to patch it with the first try, but for a different reason. Will test it and let you know if I can come up with something that works. It looks like this AndroidManifest.xml might need a manual edit before repackaging too.

**Comment by aliyuchang33** [comment]
Thank you so much. I am learning Decompiling. Hope you can achieve that.

**Comment by iamtutu** [comment]
Anyone got headsway? I am still having this issue. below my error

(virtual-python3) root@santoku-VirtualBox:/home/santoku# objection patchapk -s *****.apk 
No architecture specified. Determining it using `adb`...
Detected target device architecture as: arm64-v8a
Using latest Github gadget version: 10.6.54
Patcher will be using Gadget version: 10.6.54
Unpacking *****.apk
App already has android.permission.INTERNET
Smali not found in smali directory. This might be a multidex APK. Searching...
Traceback (most recent call last):
  File "/root/virtual-python3/bin/objection", line 11, in <module>
    sys.exit(cli())
  File "/root/virtual-python3/lib/python3.4/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/root/virtual-python3/lib/python3.4/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/root/virtual-python3/lib/python3.4/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/root/virtual-python3/lib/python3.4/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/root/virtual-python3/lib/python3.4/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/root/virtual-python3/lib/python3.4/site-packages/objection/console/cli.py", line 219, in patchapk
    patch_android_apk(**locals())
  File "/root/virtual-python3/lib/python3.4/site-packages/objection/commands/mobile_packages.py", line 155, in patch_android_apk
    patcher.inject_load_library()
  File "/root/virtual-python3/lib/python3.4/site-packages/objection/utils/patchers/android.py", line 496, in inject_load_library
    raise Exception('Unable to find smali to patch!')
Exception: Unable to find smali to patch!
Cleaning up temp files...

**Comment by leonjza** [comment]
@iamtutu That traceback looks like a different problem, where the patcher can't find the right location to inject the loadLibrary call for the frida gadget. Although this doesn't happen very often, when it does one needs to manually inspect the APK and determine the right location for it.

**Comment by burrowingsec** [comment]
So @aliyuchang33's problem looks like it might be related to resources in the APK. I've had issues with apktool extracting and building resources from APKs in the past. A way to get around any resource-related apktool issues is to use the ` -r ` flag when decompiling the APK. I've actually done this with my local copy of Objection. @leonjza if you push a fix using this method then I'm not sure the effect that might have on the ability to add ` network_security_config.xml ` functionality like you were planning on doing.

**Comment by leonjza** [comment]
Nice thank you @burrowingsec ! Can test and see what the effects are. Worst case we can just add a flag that tacks on the `-r` if needed. Much like the `--pause` can be used to make manual edits before the repackaging/signing is done.

**Comment by iamtutu** [comment]
@leonjza you were correct, apparently the app was packed using DexProtector so the dex is encrypted.

**Comment by leonjza** [comment]
@iamtutu ah perfect. Might be a useful thing to try and detect before we go ahead and try and patch it. Happy for someone to help out with this ;)

**Comment by aliyuchang33** [comment]
If i am not wrong, my apk used 360 to protect. I am trying to use some specific tools to decompile it.

**Comment by leonjza** [comment]
@aliyuchang33 I finally got to testing (and hopefully fixing) this, and it seems like I can now successfully patch and run the APK you have sent me.

I have released v1.2.5, so `pip3 install objection --upgrade`, test and let me know how it goes!

**Comment by aliyuchang33** [comment]
Yes, it works. Thanks for working so hard.
